File: 143.reorder-list.py
# @leet start
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def reorderList(self, head: Optional[ListNode]) -> None:
        slow = head
        fast = head
        while fast is not None and fast.next is not None:
            slow = slow.next
            fast = fast.next.next
        farside = slow.next
        slow.next = None
        farside = self.reverse(farside)
        answer = ListNode()
        temp = answer
        while head is not None or farside is not None:
            if head is not None:
                temp.next = head
                head = head.next
                temp = temp.next
            if farside is not None:
                temp.next = farside
                farside = farside.next
                temp = temp.next
        return answer.next

    def reverse(self, head: Optional[ListNode]) -> None:
        previous = None
        next = None
        while head:
            next = head.next
            head.next = previous
            previous = head
            head = next
        return previous

        """
        Do not return anything, modify head in-place instead.
        """


# @leet end

File: 143.reorder-list.ts
// @leet start
/**
 * Definition for singly-linked list.
 * class ListNode {
 *     val: number
 *     next: ListNode | null
 *     constructor(val?: number, next?: ListNode | null) {
 *         this.val = (val===undefined ? 0 : val)
 *         this.next = (next===undefined ? null : next)
 *     }
 * }
 */

/**
 Do not return anything, modify head in-place instead.
 */
function reorderList(head: ListNode | null): void {
	if (!head) {
		return;
	}
	const nodes = [];
	let node = head;
	while (node) {
		nodes.push(node);
		node = node.next;
	}
	let i = 0;
	let j = nodes.length - 1;
	while (i < j) {
		nodes[i].next = nodes[j];
		i++;
		if (i === j) {
			break;
		}
		nodes[j].next = nodes[i];
		j--;
	}
	nodes[i].next = null;
}
// @leet end
